### PR TITLE
Proposal of `apt-source` subcrate

### PR DESCRIPTION
**Pull request by michalfita**
## Purpose

When little over year ago I needed support for `.sources` files after Ubuntu announced move to DEB822 format as default, I couldn't find good support in Rust. I've written my own _parser_ that did the job and could handle conversion from one-liners.

Today is see for completeness `deb822-loseless` (don't like the name much) could support APT's `.sources` in DEB822 format. This PR is attempt to lie foundations for that.

## Status
This is **MVP**/**PoC** draft for sparking discussion about the sense and potential continuation. If not accepted I can personally make use of it, but I'm not willing to maintain this as separate crate for others.

I'm open to criticism 😃

## Missing `deb822-loseless` features

I found following features missing (mostly in comparison to using `serde`), list may not be complete:
* Ability to set defaults, and ignore explicitness of `Option<...>`
* Option to assume `Option<Vec<...>>` is redundant and empty `Vec<_>` would be good enough

### CODE REVIEW

**Comment by jelmer** [comment]
I'll reply in-depth when I have more time, but some quick thoughts/comments:

* Maintaining apt-sources as a subcrate of this repo makes sense, I think 

* We should probably make clear that it currently only supports the new apt-sources format; it would be nice to support the older format as well at some point

* I've renamed the repository to deb822-rs since it contains a number of deb822-related crates, deb822-lossless is just one

**Comment by michalfita** [comment]
> I've just approved running of CI, but it looks like there are some test failures.

Fixed.

> Maintaining apt-sources as a subcrate of this repo makes sense, I think

Great to hear.

> We should probably make clear that it currently only supports the new apt-sources format; it would be nice to support the older format as well at some point

I have parser for the old format, however, it switches based on file extension passed (both are intertwingled with actual file handling). I would have to be user's decision which format to use (not sure about API if loads to the same structure). Then, if it IS user's decision, it's not DEB822 format. I could bring its code here, but it's nasty regular expression processing, would require changes to work with all fields `Repository` now has, but which I just originally hold in a `HashMap<>`.

> I've renamed the repository to deb822-rs since it contains a number of deb822-related crates, deb822-lossless is just one

Make very much sense.

File: apt-sources/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "apt-sources"
+authors = ["Michał Fita <4925040+michalfita@users.noreply.github.com>"]
+edition = "2021"
+version = "0.1.40"
+license = "Apache-2.0"
+description = "A parser for APT source files (package repositories specification)"
+repository = { workspace = true }
+homepage = { workspace = true }
+keywords = ["debian", "deb822", "rfc822", "lossless", "edit"]
```
Comment (jelmer) [inline apt-sources/Cargo.toml:10]: The current crate doesn't allow format-preserving editing, so perhaps drop the lossless and edit keywords - maybe add "apt"?

File: apt-sources/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+//!  WoG/4oBsAQCEN8Z00DXagPHbwrvsY2t9BCsT+PgnSn9biobwX7bDDg==
+//!  =5NZE
+//!  -----END PGP PUBLIC KEY BLOCK-----"#;
+//!
+//! let r = text.parse::<Repositories>().unwrap();
+//! let suites = r[0].suites();
+//! assert_eq!(suites[0], "noble");
+//! ```
+//!
+//! See the ``lossless`` module (behind the ``lossless`` feature) for a more forgiving parser that
```
Comment (jelmer) [inline apt-sources/src/lib.rs:41]: This comment seems copied from another crate, but there's no actual lossless module here?

**Review by jelmer — APPROVED** [review]

**Comment by jelmer** [comment]
I think this is in a good enough shape to land. There's other improvements we can make, but we can make those after the initial version is in. What do you think?

**Comment by michalfita** [comment]
If you think it's good enough to land it the current form I'm happy to let it in. I would let me refactor my company's tool to use the library instead of my own pseudo-parser.

Few extra questions though:
* Do you have established pattern to convert lossless _structure_ (container of paragraph with accessors) into lossy structure (like in my example)? Asking for scenario, where application opens the file as lossless variant, crates lossy copy for faster and easier processing, then applies lossy → lossless for changes, then saves retaining comments, etc.
* Would it be possible to use DEB822 as Serde's serialization format, similar to JSON (I like macro building JSONs from structure, but the interoperability with Serde is awesome)? I know DEB822 has serious limitations, but maybe withing limits it could work?

**Comment by jelmer** [comment]
> If you think it's good enough to land it the current form I'm happy to let it in. I would let me refactor my company's tool to use the library instead of my own pseudo-parser.
> 
> Few extra questions though:
> 
>     * Do you have established pattern to convert lossless _structure_ (container of paragraph with accessors) into lossy structure (like in my example)? Asking for scenario, where application opens the file as lossless variant, crates lossy copy for faster and easier processing, then applies lossy → lossless for changes, then saves retaining comments, etc.

Some of the other crates implement From<> for going from one to the other.

>     * Would it be possible to use DEB822 as Serde's serialization format, similar to JSON (I like macro building JSONs from structure, but the interoperability with Serde is awesome)? I know DEB822 has serious limitations, but maybe withing limits it could work?

Yeah, I have done some work towards that but haven't gotten around to publishing it. It works - the main challenge is working out how it would intereact with the current structs.

**Review by jelmer — APPROVED** [review]

**Pull request merged**